src/Features/Core/Portable/RemoveUnusedMembers/AvoidUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,199 @@
+                            valueUsageInfo = ValueUsageInfo.Write;
+                        }
+
+                        OnSymbolUsage(memberReference.Member, valueUsageInfo);
+                    }
+                }
+
+                void AnalyzeInvocationOperation(OperationAnalysisContext operationContext)
+                {
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+                    if (IsCandidateSymbol(invocation.TargetMethod))
+                    {
+                        OnSymbolUsage(invocation.TargetMethod, ValueUsageInfo.Read);
+                    }
+                }
+
+                void OnSymbolEnd(SymbolAnalysisContext symbolEndContext)
+                {
+                    // Report diagnostics for unused candidate members.
+                    var members = ((INamedTypeSymbol)symbolEndContext.Symbol).GetMembers();
+                    foreach (var member in members)
+                    {
+                        // Check if the underlying member is neither read nor a readable reference to the member is taken.
+                        if (symbolValueUsageStateMap.TryRemove(member, out var valueUsageInfo) &&
+                            (valueUsageInfo & (ValueUsageInfo.Read | ValueUsageInfo.ReadableRef)) == 0)
+                        {
+                            Debug.Assert(IsCandidateSymbol(member));
+
+                            var option = GetCodeStyleOption(member, symbolEndContext);
+                            if (option != null && option.Value)
+                            {
+                                // Give appropriate message based on usage:
+                                //  1. No read or writes:   Type '{0}' has an unused private member '{1}' which can be removed.
+                                //  2. Only writes:         Type '{0}' has a private member '{1}' which can be removed as the value assigned to it is never used.
+
+                                var rule = (valueUsageInfo & (ValueUsageInfo.Write | ValueUsageInfo.WritableRef)) != 0 ?
+                                    _noReadsUnnecessaryWithSuggestionDescriptor :
+                                    UnnecessaryWithSuggestionDescriptor;
+
+                                var additionalLocations = member.Locations.Skip(1).ToImmutableArrayOrEmpty();
+                                var diagnostic = DiagnosticHelper.Create(
+                                    rule,
+                                    member.Locations[0],
+                                    option.Notification.Severity,
+                                    additionalLocations,
+                                    properties: ImmutableDictionary.CreateRange(new[] {
+                                        new KeyValuePair<string, string>(UnunsedMemberNameProperty, member.Name),
+                                        new KeyValuePair<string, string>(UnunsedMemberKindProperty, member.Kind.ToString()) }),
```

Comment:
1. does Kind.ToString allocate for enums.... i think it does.  We may want to consider precomputing and storing these.
2. you are allocating a temp array each time you do this.